components/shared/code/libs/LIB_nvm.c

```diff
@@ -0,0 +1,522 @@
+            if (((lib_nvm_recordHeader_S*)data.currentPtr)->discarded != SET_STATE)
+            {
+                lib_nvm_crc_t crc = 0xff;
+
+                if (((((lib_nvm_recordHeader_S*)data.currentPtr)->entryId > NVM_LAST_USED_INTERNAL_ENTRYID) &&
+                    (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId < NVM_ENTRYID_INTERNAL_COUNT)) ||
+                    (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId > NVM_TOTAL_ENTRYID_COUNT))
+                {
+                    // TODO : Handle failure
+                    // initializeNVMBlock((uint32_t)data.currentPtr); // Immediately scrap the block and start with last elements?
+                    break;
+                }
+
+                if (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId <= NVM_LAST_USED_INTERNAL_ENTRYID)
+                {
+                    crc = crc8_calculate(crc, (uint8_t*)record, lib_nvm_private_entries[id].entrySize);
+                }
+                else if (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId >= NVM_ENTRYID_INTERNAL_COUNT)
+                {
+                    crc = crc8_calculate(crc, (uint8_t*)record, lib_nvm_entries[id].entrySize);
+                }
+
+                if (crc == ((lib_nvm_recordHeader_S*)data.currentPtr)->crc)
+                {
+                    records[((lib_nvm_recordHeader_S*)data.currentPtr)->entryId].currentNonVolatileAddress_Ptr = (lib_nvm_recordHeader_S*)data.currentPtr;
+                    if (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId <= NVM_LAST_USED_INTERNAL_ENTRYID)
+                    {
+                        memcpy(lib_nvm_private_entries[id].entryRam_Ptr, (uint8_t*)record, lib_nvm_private_entries[id].entrySize);
+                    }
+                    else if (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId >= NVM_ENTRYID_INTERNAL_COUNT)
+                    {
+                        memcpy(lib_nvm_entries[id].entryRam_Ptr, (uint8_t*)record, lib_nvm_entries[id].entrySize);
+                    }
+                }
+                else
+                {
+                    storage_t disc = SET_STATE;
+                    LIB_NVM_WRITE_TO_FLASH((uint32_t)&(((lib_nvm_recordHeader_S*)data.currentPtr)->discarded),
+                                        &disc,
+                                        sizeof(storage_t));
+                    recordLog.totalFailedCrc++;
+                }
+            }
+
+            if (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId <= NVM_LAST_USED_INTERNAL_ENTRYID)
+            {
+                data.currentPtr += lib_nvm_private_entries[id].entrySize / sizeof(storage_t);
+            }
+            else if (((lib_nvm_recordHeader_S*)data.currentPtr)->entryId >= NVM_ENTRYID_INTERNAL_COUNT)
+            {
+                data.currentPtr += lib_nvm_entries[id].entrySize / sizeof(storage_t);
+            }
+
+            data.currentPtr += sizeof(lib_nvm_recordHeader_S) / sizeof(storage_t);
```

Comment:
Logically incrementing by header should be first, and add entrySize to the summation

Reply:
needs to be first